### PR TITLE
Exercise5

### PR DESCRIPTION
**Pull request by MatzeK105**
Software tests are in progress.

We will review @abteilung6.

Review: https://github.com/Systems-Development-and-Frameworks/abteilung6/pull/7#pullrequestreview-564793434

### CODE REVIEW

File: backend/src/datasource/post-datasource.js [hunk 1]
```diff
@@ -80,4 +84,16 @@ export class PostDatasource extends DataSource {
...
     }
+
+    async getPosts() {
+        const {data} = await this.query(gql`{ posts { title }}`)
+
+        return data.posts
+    }
+
+    async writePosts() {
+        const {data} = await this.mutation(gql`{ write (post:{title:"testPost"}) { title } }`)
```
Comment (OliKue) [inline backend/src/datasource/post-datasource.js:93]: Keep in mind to change this before merging 👍

File: exercises/README.md [hunk 2]
```diff
@@ -4,5 +4,6 @@
 * [Exercise #2](./2/README.md)
 * [Exercise #3](./3/README.md)
 * [Exercise #4](./4/README.md)
+* [Exercise #5](./5/README.md)
```
Comment (OliKue) [inline exercises/README.md:8]: I think this should not appear here when you rebase on the homework-repo but im not sure...

File: exercises/5/README.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
+# Exercise #5
```
Comment (OliKue) [inline exercises/5/README.md:1]: I think this should not appear here when you rebase on the homework-repo but im not sure...

File: exercises/4/README.md [hunk 4]
```diff
@@ -13,35 +13,33 @@ Extend exercise [#1](../1), [#2](../2) and [#3](../3) with the new [objectives](
 
 ## Instructions
 
-1. Rebase your `main` branch on top of `homework/main`.
```
Comment (OliKue) [inline exercises/4/README.md:16]: I think this should not appear here when you rebase on the homework-repo but im not sure...

File: backend/src/typeDefs.js [hunk 5]
```diff
@@ -6,27 +6,25 @@ type Post {
   id: ID!
```
Comment (OliKue) [inline backend/src/typeDefs.js]: Why did you change the names?

File: backend/src/test/integration.spec.js [hunk 6]
```diff
@@ -1,536 +1,538 @@
...
+//                   ],
+//                 },
+//               },
+//             ],
+//           },
+//         },
+//       ],
+//     });
+//   });
+// });
```
Comment (OliKue) [inline backend/src/test/integration.spec.js:538]: Keep in mind to bring those test back alive :D

File: backend/src/datasource/permissions.js [hunk 7]
```diff
@@ -16,7 +16,7 @@ const isAuthenticatedUser = rule({cache: 'contextual'})(
 
 export const permissions = shield({
   Query: {
-    users: allow,
+    people: allow,
```
Comment (OliKue) [inline backend/src/datasource/permissions.js:19]: Why did you change the names?

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+git-crypt-key
```
Comment (OliKue) [inline .gitignore:1]: ⭐ For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.

**Review by OliKue — COMMENTED** [review]
On your current draft PR:

❕ ⭐ For choosing a scenario and writing installation instructions in README.md.

❕ ⭐ For explaining WHY you have chosen the scenario in the README.md.

⭐ For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.

❕ ⭐ ❕ ⭐ Your software tests are free of side effects. (currently not all Tests are working // aktive)

⭐ Your database is never left in an invalid state.

File: backend/.env.test [hunk 9]
```diff
@@ -1 +1 @@
-JWT_SECRET = Te5tK3Y!
\ No newline at end of file
+JWT_SECRET = Pa$$w0rd
```
Comment (roschaefer) [inline backend/.env.test:1]: Seems not to be a secret as it's in `.env.test` and `.env` is encrypted. :+1: 
> :star: For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.

File: README.md [hunk 10]
```diff
@@ -20,16 +20,27 @@ You need NodeJS, NPM, VueJS installed on your computer. To run the application,
 
 ### Backend
 
+Why GraphCMS?
+ - GraphCMS is a headless CMS where you can store and get data from
+ - You don't need to set up a database
+ - Your data is hosted without the need of a self-hosted server
+ - The hosting is free (:
```
Comment (roschaefer) [inline README.md:27]: > :star: For explaining *WHY* you have chosen the scenario in the `README.md`.

File: README.md [hunk 11]
```diff
@@ -20,16 +20,27 @@ You need NodeJS, NPM, VueJS installed on your computer. To run the application,
...
-6. Run `npm run dev` to start the application server.
-7. Open `localhost:4000` in your web browser to view the application.
+2. (This course only !) Decrypt the `.env` file with `git-crypt-key` provided by the mentors.
+3. Create a `.env`file
+4. Add a JWT Secret similar to this `JWT_SECRET = Your_Secret`
+5. Add a GRAPH_CMS_API_TOKEN that is provided by GraphCMS similar to this `GRAPH_CMS_API_TOKEN = Your_Token`
+6. Add a GRAPH_CMS_ENDPOINT that is provided by GraphCMS similar to this `GRAPH_CMS_ENDPOINT = Your_Endpoint_Url`
+7. Save the `.env`file
+8. Run `npm install`.
+9. Run `npm run dev` to start the application server.
```
Comment (roschaefer) [inline README.md:37]: What about `npm run test`?

File: backend/src/datasource/permissions.js (hunk 7 above)
Comment (roschaefer) [reply backend/src/datasource/permissions.js:19]: "User" is a reserved word on GraphCMS. You cannot call one of your own types `User`.

File: backend/src/datasource/post-darasource-graphcms.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+const {DataSource} = require('apollo-datasource');
+
+export class PostDatasourceGraphCMS extends DataSource {
+    userVoted(votes, userId) {
+        return votes.find(vote => vote.person.id == userId) ? true : false
+    }
+
+    getVotes(post_id) {
+        let post = this.posts.find(post => post.id == post_id)
```
Comment (roschaefer) [inline backend/src/datasource/post-darasource-graphcms.js:9]: Where does `this.posts` come from? :thinking:

Comment (roschaefer) [reply backend/src/datasource/post-darasource-graphcms.js:9]: I think this is dead code right? :arrow_right: :wastebasket:

File: backend/src/datasource/post-datasource.js [hunk 13]
```diff
@@ -80,4 +82,16 @@ export class PostDatasource extends DataSource {
...
+
+        return data.posts
+    }
```
Comment (roschaefer) [inline backend/src/datasource/post-datasource.js:96]:
```
$ git grep 'writePost'
src/datasource/post-datasource.js:    async writePosts() {
```
So this is dead code, too. Please delete :wastebasket:

File: backend/src/resolvers.js [hunk 14]
```diff
@@ -1,22 +1,301 @@
...
+        document: createPostMutation
+      });
+      if (errors) throw new Error(errors.map((e) => e.message).join("\n"));
+
+      return data.createPost;
+    },
+    delete: async (parent, args, context, info) => {
+      const deleteMutation = gql`
+        mutation {
+          deletePost(where: {id: "${args.id}"}) {
```
Comment (roschaefer) [inline backend/src/resolvers.js:74]: Use query arguments:
```gql
mutation($id: ID!) {
  deletePost(where: {id: $id}) { }
}
```

File: backend/src/resolvers.js [hunk 15]
```diff
@@ -1,22 +1,301 @@
...
+      if (!person) return "Email or Password incorrect.";
+
+      return context.dataSources.udbg.checkPassword(
+        person.id,
+        args.password,
+        person.password
+      );
+    },
+    vote: async (parent, args, context, info) => {
+      if (args.voteValue != 1 && args.voteValue != -1)
```
Comment (roschaefer) [inline backend/src/resolvers.js:146]:
```js
       if ( ! [-1, 1].includes(args.voteValue))
```

File: backend/src/resolvers.js [hunk 16]
```diff
@@ -1,22 +1,301 @@
...
+        } else {
+          // create new vote
+          voteMutation = gql`
+              mutation {
+                createVote(data: {value: ${args.voteValue}, person: {connect: {id: "${context.userData.userId}"}}, post: {connect: {id: "${args.postId}"}}}) {
+                  id
+                  value
+                }
+              }
+            `;
```
Comment (roschaefer) [inline backend/src/resolvers.js:208]: DRY this! The two `else` branches can be merged.

File: backend/src/resolvers.js [hunk 17]
```diff
@@ -1,22 +1,301 @@
...
+        }
+        `;
+
+      const response = await executor({ document: authorQuery });
+      if (response.errors)
+        throw new Error(response.errors.map((e) => e.message).join("\n"));
+
+      return response.data.person;
+    },
+    voteResult: async (parent, args, context) => {
```
Comment (roschaefer) [inline backend/src/resolvers.js:257]: If you would use schema stitching this would be super easy to implement without additional call to the graphCMS endpoint. This implementation is adding N+1 queries to execution (terrible performance on complex queries).

File: backend/src/schema.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
+import { stitchSchemas } from '@graphql-tools/stitch';
+import { applyMiddleware } from 'graphql-middleware';
+import { typeDefs } from './typeDefs';
+import { permissions } from './datasource/permissions';
+import Resolvers from './resolvers';
+import GraphCmsSchema, { executor } from './graphCms/schema';
+
+export default async () => {
+  const graphCmsSchema = await GraphCmsSchema();
+  const resolvers = Resolvers({ schema: graphCmsSchema, executor });
```
Comment (roschaefer) [inline backend/src/schema.js:10]: Here you can remove certain fields from the resulting schema: https://github.com/Systems-Development-and-Frameworks/homework/blob/demo/demo/backend/src/schema.js#L11

File: backend/src/test/resolvers.spec.js [hunk 19]
```diff
@@ -0,0 +1,590 @@
...
+jest.mock("../graphCms/schema");
+
+let reqMock;
+let resMock;
+let query;
+let mutate;
+let contextMock = () => context({ req: reqMock, res: resMock });
+
+beforeEach(async () => {
+  const server = await Server({ context: () => contextMock });
```
Comment (roschaefer) [inline backend/src/test/resolvers.spec.js:16]: That's why the tests are taking so long. You're re-initializing the entire server for ever test case. Because the only data that is being changed in tests is `dataSources`, `context` and your mock `GraphCmsSchema`, you don't have to re-initialize the whole server, just the parts that change.

File: backend/src/test/resolvers.spec.js [hunk 20]
```diff
@@ -0,0 +1,590 @@
...
+      });
+    });
+
+    describe("Post properties", () => {
+      it("Returns all properties", async () => {
+        let res = await query({ query: postQuery });
+
+        // voteResult = 12 because votes returns 2 instances with value = 6 each
+        expect(res.data).toEqual({
+          posts: [
```
Comment (roschaefer) [inline backend/src/test/resolvers.spec.js:78]: Does `addMocksToSchema` always return two objects for an array?

File: backend/src/test/resolvers.spec.js [hunk 21]
```diff
@@ -0,0 +1,590 @@
...
+      const signupMut = gql`
+        mutation {
+          signup(name: "Peter", email: "mail@mail.de", password: "pass")
+        }
+      `;
+      let res = await mutate({ mutation: signupMut });
+      expect(res.data).toEqual({
+        signup: "Password invalid. Choose another password.",
+      });
+    });
```
Comment (roschaefer) [inline backend/src/test/resolvers.spec.js:563]: In order to test if the remote schema would be called, you could inject `executor` into your schema from within your test cases. Your `executor` would be a mock function and you could ask:
```js
expect(executor).toHaveBeenCalled()
```
https://jestjs.io/docs/en/expect#tohavebeencalled

File: exercises/4/README.md (hunk 4 above)
Comment (roschaefer) [reply exercises/4/README.md:16]: Yes!

File: backend/src/resolvers.js [hunk 22]
```diff
@@ -1,22 +1,301 @@
...
+        args.password
+      );
+
+      if (!passwordHash) return "Password invalid. Choose another password.";
+
+      const createUser = gql`
+        mutation {
+          createPerson(data: {name: "${
+            args.name
+          }", email: "${args.email.toLowerCase()}", password: "${passwordHash}"}) {
```
Comment (roschaefer) [inline backend/src/resolvers.js:113]: The field `password` should be called `passwordHash`. Better naming, right?

**Review by roschaefer — COMMENTED** [review]
well done

Hey Team `SCP-Foundation` you received 8/10 :star: in this submission. Well done.

Could you explain why you chose not to use `delegateToSchema`?

There are two major issues with your code. As you can see, the client is able
to call the subschema directly, bypassing all authorization.

Also, you encrypted your `backend/.env` file with `git-crypt` but you didn't add
your API keys there. From the instructions:
> 2. If you require API keys (e.g. [Remote API](#using-a-remote-graphqlrest-api) scenario), put them into a `.env` file and encrypt this file with [git-crypt](https://github.com/AGWA/git-crypt).

That's why I have to deduce one :star: here. I am just not able to reproduce your setup, because my GraphCMS schema is obviously different from yours. At least I could (with changes to the code) run a `signup` mutation.

Apart from that I strongly suggest that you learn how to use GraphQL query variables. Your code is riddled with unsafe string interpolation. GraphQL has its own syntax for variables in a query:
https://graphql.org/learn/queries/#variables

Finally, you have a N+1 problem in your resolvers. With schema delegation, your
local gateway schema would request all data required from the remote schema.
In terms of performance, your implementation is similar to calling a REST
endpoint. You can do this in a much better way.

:star: For choosing a scenario and writing installation instructions in `README.md`.

:star: For explaining *WHY* you have chosen the scenario in the `README.md`.

:star: For not committing sensitive secrets (e.g. API keys) unencrypted to the
repository.

:star: :x: Data created in mutations are persisted and your queries
and mutations still work as expected.

:star: :star: Your software tests are free of side effects.

:star: Your database is never left in an invalid state.

:x: You deny the client to call mutations and queries of your subschema
directly.




:star: For requesting a review and reviewing another team's PR.

See further suggestions below.
